scripts/broadcast-pndcp-discovery.nse

```diff
@@ -0,0 +1,449 @@
+              --   pndcpParseDhcpBlock(suboption, blockdata, identify_response.DHCP)
+
+              else
+                stdnse.debug1("Encountered unknown DCP block: Postion: %u, Option: %u, Suboption: %u, Blocklen: %u, Data: %s, Next position: %u", block_pos, option, suboption, blocklen, stdnse.tohex(blockdata), pos)
+              end
+
+            end
+
+            responses[get_mac_addr(f.mac_src)] = identify_response
+          end
+        end
+      end
+    end
+  end
+  condvar("signal")
+end
+
+-- Sends a Profinet DCP identify request.
+-- @param interface Network interface to send on.
+local pndcpIdentify = function(interface, responseDelay)
+  local sock = nmap.new_dnet()
+  sock:ethernet_open(interface.device)
+
+  -- Wait for the listener threads to start up and open the receiving socket
+  stdnse.sleep(0.1)
+
+  -- Build DCP probe
+  local pn_dcp_identify = string.pack(">I2 B B I4 I2 I2 B B I2",
+    0xfefe,     -- Frame ID
+    0x05,       -- Service ID: 5 = Identify
+    0x00,       -- Service Type: 0 = Request
+    0x00000001, -- Xid (transaction ID): 1
+    responseDelay, -- Response delay
+    0x0004,     -- DCP Data length (length of following data)
+
+    0xff,       -- Option: 0xff = all
+    0xff,       -- Suboption: 0xff = all
+    0x0000      -- Length of following block data: 0
+  )
+
+  -- Build probe ethernet frame
+  local probe = packet.Frame:new()
+  probe.mac_src = interface.mac
+  probe.mac_dst = packet.mactobin("01:0e:cf:00:00:00")
+  probe.ether_type = string.pack(">I2", 0x8892)
+  probe.buf = tostring(pn_dcp_identify)
+  probe:build_ether_frame()
+
+  sock:ethernet_send(probe.frame_buf)
+  sock:ethernet_close()
+end
+
+
+action = function(host, port)
+  local responses, interfaces, lthreads = {}, {}, {}
+
+  local timeout = stdnse.parse_timespec(stdnse.get_script_args(SCRIPT_NAME .. ".timeout"))
+  local interface = stdnse.get_script_args(SCRIPT_NAME .. ".interface")
+
+  -- Calculate response delay values as per spec
+  -- See DIN/EN 61158-6-10 chapter 4.3.1.3.5
+  timeout = (timeout or 1)
+  local responseDelay = (timeout - 1) * 100
+
+  -- clamp the response delay to allowed values per spec
+  if responseDelay <= 0 then
+    responseDelay = 1
+  elseif responseDelay > 6400 then
+    responseDelay = 6400
+  end
+
+  -- Check the interface
+  interface = interface or nmap.get_interface()
+  if interface then
+    -- Get the interface information
+    interface = nmap.get_interface_info(interface)
+    if not interface then
+      return stdnse.format_output(false, ("Failed to retrieve %s interface information."):format(interface))
+    end
+    interfaces = {interface}
+    stdnse.debug1("Will use %s interface.", interface.shortname)
+  else
+    local ifacelist = nmap.list_interfaces()
+    for _, iface in ipairs(ifacelist) do
+      -- Match all ethernet interfaces
+      if iface.address and iface.link == "ethernet" and
+        iface.address:match("%d+%.%d+%.%d+%.%d+") then
```

Comment:
No, it doesn't need an IP address, you are right.

I am trying this right now, but I am getting an error when I call `ethernet_open`:
```
bad argument #1 to 'ethernet_open' (device is not valid ethernet interface)
```

The first interface it complains about is `enp5s0` which doesn't even have a link and is marked `DOWN`, yet `iface.up` contains `up` when I print it:
```
2: enp5s0: <NO-CARRIER,BROADCAST,MULTICAST,UP> mtu 1500 qdisc mq state DOWN group default qlen 1000
    link/ether a0:36:9f:c8:cb:c3 brd ff:ff:ff:ff:ff:ff
```
```
NSE: [broadcast-pndcp-discovery] enp5s0: up
```

Any ideas why?

Reply:
The interface is actually up (`NO-CARRIER,BROADCAST,MULTICAST,UP`) but has no carrier (no peer connected to it. You may want to catch that error? I have not been able to find another interface flag that would tell you if the interface is running.

Reply:
Thanks @p-l-, at least I understand the error message now. The interface is up, but there's no cable plugged in, so it sees no carrier. I am going to catch that error and then don't use this interface.

Funnily enough, right now everything works because I changed my network setup and don't have a running interface without a link right now. I guess I was "lucky" the first time to even run into this issue.

Reply:
Ha! So a `try`/`catch` didn't work. I grepped through the nmap code to try and find out how `ethernet_open` works and when it throws this error so that I can try and check for it mysel before I call it.

Turns out it just calls the C-equivalent of `nmap.get_interface_info`. This is what I'm doing now as well: It returns `false` when it fails. And it appears to fail when there's no carrier.